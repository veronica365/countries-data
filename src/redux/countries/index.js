import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';
import { API } from '../../constants';

const findItem = (id, countries) => countries.filter((country) => String(country.id) === id);
const searchItem = (param, countries) => countries.filter(
  (country) => String(country.id) === param
      || String(country.name).toLowerCase().includes(String(param).toLowerCase())
      || String(country.currency).toLowerCase().includes(String(param).toLowerCase())
      || String(country.capital).toLowerCase().includes(String(param).toLowerCase())
      || String(country.region).toLowerCase().includes(String(param).toLowerCase())
      || String(country.currency_name).toLowerCase().includes(String(param).toLowerCase())
      || String(country.subregion).toLowerCase().includes(String(param).toLowerCase()),
);

export const getData = createAsyncThunk('get', async (_, thunkAPI) => {
  try {
    const { countries } = thunkAPI.getState().countries;
    if (countries.length) return countries;
    const { data } = await axios.get(`${API}`);
    return data;
  } catch (error) {
    return thunkAPI.rejectWithValue('Something went wrong');
  }
});
export const findCountry = createAsyncThunk('find', async (id, thunkAPI) => {
  try {
    const { countries } = thunkAPI.getState().countries;
    if (countries.length) return findItem(id, countries);
    const { data } = await axios.get(`${API}`);
    return findItem(id, data);
  } catch (error) {
    return thunkAPI.rejectWithValue('Something went wrong');
  }
});
export const searchCountry = createAsyncThunk(
  'search',
  async (param, thunkAPI) => {
    try {
      const { countries } = thunkAPI.getState().countries;
      if (countries.length) return searchItem(param, countries);
      const { data } = await axios.get(`${API}`);
      return searchItem(param, data);
    } catch (error) {
      return thunkAPI.rejectWithValue('Something went wrong');
    }
  },
);

const countrieslice = createSlice({
  name: 'countries',
  initialState: {
    countries: [],
    countriesBySearch: [],
    countryById: {},
    isLoading: false,
    error: null,
  },
  reducers: {
    setCountryById: (state, action) => ({
      ...state,
      countryById: action.payload,
      countriesBySearch: [],
      isLoading: false,
    }),
  },
  extraReducers: (builder) => {
    builder.addCase(getData.pending, (state) => ({
      ...state,
      isLoading: true,
    }));
    builder.addCase(getData.fulfilled, (state, action) => ({
      ...state,
      countries: action.payload,
      isLoading: false,
    }));
    builder.addCase(getData.rejected, (state, action) => ({
      ...state,
      error: action.payload,
      isLoading: false,
    }));
    builder.addCase(findCountry.pending, (state) => ({
      ...state,
      isLoading: true,
    }));
    builder.addCase(findCountry.fulfilled, (state, action) => ({
      ...state,
      countryById: action.payload[0] || {},
      isLoading: false,
    }));
    builder.addCase(findCountry.rejected, (state, action) => ({
      ...state,
      error: action.payload,
      isLoading: false,
    }));
    builder.addCase(searchCountry.pending, (state) => ({
      ...state,
      isLoading: true,
    }));
    builder.addCase(searchCountry.fulfilled, (state, action) => ({
      ...state,
      countriesBySearch: action.payload,
      isLoading: false,
    }));
    builder.addCase(searchCountry.rejected, (state, action) => ({
      ...state,
      error: action.payload,
      isLoading: false,
    }));
  },
});

export const { setCountryById } = countrieslice.actions;

export default countrieslice.reducer;
