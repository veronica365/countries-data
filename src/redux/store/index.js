import { configureStore } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import countryReducer from '../countries';

const store = configureStore({
  reducer: {
    countries: countryReducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(logger),
  devTools: process.env.NODE_ENV !== 'production',
});

export default store;
