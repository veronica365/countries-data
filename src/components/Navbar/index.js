import { useEffect, useRef } from 'react';
import { useDispatch } from 'react-redux';
import { Link, useLocation, useNavigate } from 'react-router-dom';
import { getData, searchCountry, setCountryById } from '../../redux/countries';

export default function Navbar() {
  const oldSearch = useRef(null);
  const dispatch = useDispatch();
  const location = useLocation().pathname.split('/')[2];
  const navigate = useNavigate();
  useEffect(() => {
    dispatch(getData());
  }, [dispatch]);

  function delayedSearch({ target: { value } }) {
    oldSearch.current = value;
    setTimeout(async () => {
      if (oldSearch.current === value) return dispatch(searchCountry(value));
      return '';
    }, 500);
  }
  const clickHandler = (event) => {
    event.preventDefault();
    dispatch(setCountryById({}));
    navigate('/');
  };
  return (
    <header className="App-header">
      <div className="brand">
        {location && (
          <Link to="/" onClick={clickHandler}>
            <i className="fa fa-arrow-left" />
          </Link>
        )}
        <img src="/img/countries.png" alt="logo" />
        Country Data
      </div>
      <div className="search">
        <input
          type="text"
          onChange={delayedSearch}
          placeholder="Search by country, currency, city, region, etc"
        />
        <button type="button">
          <i className="fa fa-search" />
        </button>
      </div>
    </header>
  );
}
