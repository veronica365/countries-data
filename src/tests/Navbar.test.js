import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { render, fireEvent } from '@testing-library/react';
import axios from 'axios';
import Navbar from '../components/Navbar';
import store from '../redux/store';
import { getData, searchCountry } from '../redux/countries';
import { muchCountryData } from './__Mocks__/countries';

jest.mock('axios');

describe('Navbar Component', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    jest.resetModules();
    jest.restoreAllMocks();
    axios.get.mockResolvedValue({
      data: [muchCountryData],
    });
  });
  it('should render and navbar and fetch countries', () => {
    const dispatch = jest.spyOn(store, 'dispatch');
    const { container } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Navbar />
        </BrowserRouter>
      </Provider>,
    );
    expect(container).toMatchSnapshot();
    expect(axios.get).toHaveBeenCalledTimes(1);
    expect(axios.get).toHaveBeenCalledWith(
      'https://raw.githubusercontent.com/dr5hn/countries-states-cities-database/master/countries.json',
    );
    expect(dispatch).toHaveBeenCalled();
  });
  it('Should be able to fetch data for all countries', () => {
    const { container } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Navbar />
        </BrowserRouter>
      </Provider>,
    );

    expect(container).toMatchSnapshot();
    expect(store.getState().countries.countries.length).toEqual(1);
  });
  it('Should be able to search for a country from the data', async () => {
    const { container } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Navbar />
        </BrowserRouter>
      </Provider>,
    );

    expect(container).toMatchSnapshot();
    fireEvent.input(container.querySelector('input'), {
      target: {
        value: 'uganda',
      },
    });
    expect(store.getState().countries.countries.length).toEqual(1);
    expect(store.getState().countries.countriesBySearch.length).toEqual(0);
    fireEvent.input(container.querySelector('input'), {
      target: {
        value: 'a',
      },
    });
    await store.dispatch(searchCountry('a'));
    expect(store.getState().countries.countriesBySearch.length).toEqual(1);
  });
  it('Should be able to search data for all countries that much a query param', async () => {
    await store.dispatch(searchCountry('a'));
    expect(store.getState().countries.countriesBySearch.length).toEqual(1);
  });
  it('Should be able to fetch data for all countries', async () => {
    await store.dispatch(getData());
    expect(store.getState().countries.countries.length).toEqual(1);
  });
});
