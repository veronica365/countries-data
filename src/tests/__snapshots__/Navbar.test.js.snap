// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar Component Should be able to fetch data for all countries 1`] = `
<div>
  <header
    class="App-header"
  >
    <div
      class="brand"
    >
      <img
        alt="logo"
        src="/img/countries.png"
      />
      Country Data
    </div>
    <div
      class="search"
    >
      <input
        placeholder="Search by country, currency, city, region, etc"
        type="text"
      />
      <button
        type="button"
      >
        <i
          class="fa fa-search"
        />
      </button>
    </div>
  </header>
</div>
`;

exports[`Navbar Component Should be able to search for a country from the data 1`] = `
<div>
  <header
    class="App-header"
  >
    <div
      class="brand"
    >
      <img
        alt="logo"
        src="/img/countries.png"
      />
      Country Data
    </div>
    <div
      class="search"
    >
      <input
        placeholder="Search by country, currency, city, region, etc"
        type="text"
      />
      <button
        type="button"
      >
        <i
          class="fa fa-search"
        />
      </button>
    </div>
  </header>
</div>
`;

exports[`Navbar Component should render and navbar and fetch countries 1`] = `
<div>
  <header
    class="App-header"
  >
    <div
      class="brand"
    >
      <img
        alt="logo"
        src="/img/countries.png"
      />
      Country Data
    </div>
    <div
      class="search"
    >
      <input
        placeholder="Search by country, currency, city, region, etc"
        type="text"
      />
      <button
        type="button"
      >
        <i
          class="fa fa-search"
        />
      </button>
    </div>
  </header>
</div>
`;
